### ISSUE TITLE
Fails to detect endianes and generate compose sequences when cross compiling

### ISSUE THREAD

**Issue by oreo639** [ibus/ibus#2479]
When trying to cross compile ibus 1.5.28 on Void Linux using:
```
./configure --prefix=/usr --sysconfdir=/etc --sbindir=/usr/bin --bindir=/usr/bin --mandir=/usr/share/man --infodir=/usr/share/info --localstatedir=/var --host=x86_64-unknown-linux-gnu --build=x86_64-unknown-linux-gnu '--libdir=${exec_prefix}/lib64' --host=aarch64-linux-gnu --with-sysroot=/usr/aarch64-linux-gnu --with-libtool-sysroot=/usr/aarch64-linux-gnu --enable-ui --enable-gtk3 --enable-gtk4 --enable-memconf --enable-dconf --enable-wayland --enable-xim --disable-tests --disable-schemas-compile --disable-systemd-services --with-python=/usr/bin/python3 --disable-python2 --disable-emoji-dict --disable-unicode-dict --enable-introspection --enable-vala --enable-setup
```

It first fails at:
```
checking build system endianness... configure: error: in `/home/oreo639/git/IPR/ibus':
configure: error: cannot run test program while cross compiling
See `config.log' for more details
```

This is because here:
https://github.com/ibus/ibus/blob/main/configure.ac#L157-L178

`AC_RUN_IFELSE()` is used instead of `AC_COMPILE_IFELSE()`. Replacing it with `AC_COMPILE_IFELSE()` allows the check to complete sucessfully.

After this, a second build failure occurs with:
```
INFO: Reading ./ibusmarshalers.list...
  CCLD     gen-internal-compose-table
  GEN      compose/sequences-little-endian
/bin/sh: line 2: ./gen-internal-compose-table: cannot execute binary file: Exec format error
make[2]: *** [Makefile:1717: compose/sequences-little-endian] Error 126
make[2]: Leaving directory '/builddir/ibus-1.5.28/src'
make[1]: *** [Makefile:697: all-recursive] Error 1
make[1]: Leaving directory '/builddir/ibus-1.5.28'
make: *** [Makefile:604: all] Error 2
```
As automake tries to generate `sequences-little-endian` and `sequences-big-endian` despite those files already existing.

Idk how best to fix this, however just for reference, the following patch did allow me to workaround it and sucessfully cross compile ibus:
```
diff --git a/src/Makefile.am b/src/Makefile.am
index 426376dd..fd6ed429 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -269,10 +269,7 @@ ibus.gresources.xml: ibus.gresources.xml.in
 	mv $@.tmp $@
 
 compose/sequences-$(ENDIAN)-endian: gen-internal-compose-table
-	$(AM_V_GEN) $(MKDIR_P) $(builddir)/compose && \
-	$(builddir)/gen-internal-compose-table && \
-	mv $(builddir)/sequences-big-endian $(builddir)/compose && \
-	mv $(builddir)/sequences-little-endian $(builddir)/compose
+	$(AM_V_GEN) $(MKDIR_P) $(builddir)/compose
 
 ibusresources.h: ibus.gresources.xml compose/sequences-$(ENDIAN)-endian
 	$(AM_V_GEN) $(GLIB_COMPILE_RESOURCES) --generate-header --internal \
```

Ofc this would need to be properly fixed either by preventing the files from being built if they already exist or by compiling `gen-internal-compose-table` for the build system rather than the host system.

**Comment by fujiwarat** [comment]
> /bin/sh: line 2: ./gen-internal-compose-table: cannot execute binary file: Exec format error

I don't understand why this error happens. It happens because the build system of making gen-internal-compose-table and the execution system of running gen-internal-compose-table have the different architectures?

How about the following change?
```
diff --git a/src/Makefile.am b/src/Makefile.am
index c2e8daa6..81c555f8 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -269,10 +269,12 @@ ibus.gresources.xml: ibus.gresources.xml.in
        mv $@.tmp $@
 
 compose/sequences-$(ENDIAN)-endian: gen-internal-compose-table
-       $(AM_V_GEN) $(MKDIR_P) $(builddir)/compose && \
-       $(builddir)/gen-internal-compose-table && \
-       mv $(builddir)/sequences-big-endian $(builddir)/compose && \
-       mv $(builddir)/sequences-little-endian $(builddir)/compose
+       $(AM_V_GEN) if test ! -f $(builddir)/compose/sequences-$(ENDIAN)-endian; then \
+           $(MKDIR_P) $(builddir)/compose && \
+           $(builddir)/gen-internal-compose-table && \
+           mv $(builddir)/sequences-big-endian $(builddir)/compose && \
+           mv $(builddir)/sequences-little-endian $(builddir)/compose; \
+       fi
 
 ibusresources.h: ibus.gresources.xml compose/sequences-$(ENDIAN)-endian
        $(AM_V_GEN) $(GLIB_COMPILE_RESOURCES) --generate-header --internal \
```

**Comment by oreo639** [comment]
> I don't understand why this error happens. It happens because the build system of making gen-internal-compose-table and the execution system of running gen-internal-compose-table have the different architectures?

Yes, I am cross-compiling. In this case `--build` (the system I am compiling on) is x86_64 and `--host` (the system I am compiling for) is aarch64. You might not notice this issue if you are using binfmt to autorun qemu I guess.

> How about the following change?

Yes, that does work.

**Comment by fujiwarat** [comment]
OK, I think there are two issues. One is the cross-compile issue and another is the dependency issue that gen-internal-compose-table is built even if compose/sequences-$(ENDIAN)-endian exists. Currently I have no idea about this kind of the weak dependencies with the Makefile rules and adding "if" sentence could be the workaround as my suggestion.

On the other hand, I prefer your patch and let's think it.

**Comment by alyssais** [comment]
For me this has been resolved with 630f3f73f62a121abb1af5b215d193aec4e0053c.
